### PR TITLE
Rewrite the onUpdate and delete method of pod 

### PR DESCRIPTION
**Pull request by dxsup**
## Description

1. Replace PodInfo with k8sPodInfo in the globalPodInfo. This helps reduce the complexity of the data structure of pods. 
2. Only delete the desired fields of pods when delayed deleting.
3. Rewrite the `onUpdate` method to avoid the delayed deleting requests removing everything.

## Related Issue

#239 

## How Has This Been Tested?

Now the test case that failed before is passed.
```
=== RUN   TestUpdateDelayDelete
    pod_watch_test.go:218: Found container [192.168.136.210:9001]
    pod_watch_test.go:228: Found container [192.168.136.210:9001]
--- PASS: TestUpdateDelayDelete (0.61s)
PASS
```

### CODE REVIEW

File: collector/metadata/kubernetes/pod_watch.go [hunk 1]
```diff
@@ -248,22 +240,115 @@ func OnUpdate(objOld interface{}, objNew interface{}) {
...
+
+	// Check the containers' ID
+	newContainerIds := make([]string, 0)
+	for _, containerStatus := range oldPod.Status.ContainerStatuses {
+		shortenContainerId := TruncateContainerId(containerStatus.ContainerID)
+		if shortenContainerId == "" {
+			continue
+		}
+		newContainerIds = append(newContainerIds, shortenContainerId)
+	}
```
Comment (NeJan2020) [inline collector/metadata/kubernetes/pod_watch.go:275]: Here we need to search newContainerIds  from newPod.status.ContainerStatues instead of `oldPod`

Comment (dxsup) [reply collector/metadata/kubernetes/pod_watch.go:275]: Sorry, it's a typo.

File: collector/metadata/kubernetes/pod_watch.go [hunk 2]
```diff
@@ -248,22 +240,115 @@ func OnUpdate(objOld interface{}, objNew interface{}) {
...
+		}
+	}
+
+	portsCompare := compare.NewInt32Slice(oldCachePod.Ports, newPorts)
+	portsCompare.Compare()
+	deletedPodInfo.ports = portsCompare.GetRemovedElements()
+
+	hostPortsCompare := compare.NewInt32Slice(oldCachePod.HostPorts, newHostPorts)
+	hostPortsCompare.Compare()
+	deletedPodInfo.hostPorts = hostPortsCompare.GetRemovedElements()
```
Comment (NeJan2020) [inline collector/metadata/kubernetes/pod_watch.go:303]: When the pod IP changes but the port and hostPort remain unchanged, the hostIpPort and podIpPort information will not be updated normally.

But maybe the pod update could not modify the pod IP or pod Name.

Comment (NeJan2020) [reply collector/metadata/kubernetes/pod_watch.go:303]: I have updated the testcase about this problem

Comment (dxsup) [reply collector/metadata/kubernetes/pod_watch.go:303]: Actually, it could. I' ll fix this   issue.

**Review by NeJan2020 — APPROVED** [review]

**Review by NeJan2020 — APPROVED** [review]

**Pull request merged**